### PR TITLE
Make docroot and server name configurable.

### LINKED ISSUE #6: Configurable DocumentRoot
via environment variable

TBA

### PR DESCRIPTION
**Pull request by slashrsm**
Based on discussion in https://github.com/drupal-docker/php/issues/46.

Fixes #6.

### CODE REVIEW

File: mainline-alpine/Dockerfile [hunk 1]
```diff
@@ -11,3 +11,6 @@ RUN addgroup -g 82 -S www-data \
 
 COPY nginx.conf /etc/nginx/nginx.conf
 COPY drupal.conf /etc/nginx/conf.d/default.conf
+
+COPY entrypoint.sh /usr/local/bin/
+CMD "entrypoint.sh"
```
Comment (zaporylie) [inline mainline-alpine/Dockerfile]: Should use ENTRYPOINT instruction

File: mainline-alpine/drupal.conf [hunk 2]
```diff
@@ -1,5 +1,5 @@
 server {
-    server_name _;
+    server_name SERVER_NAME;
```
Comment (zaporylie) [inline mainline-alpine/drupal.conf]: Why do we need it?

File: mainline-alpine/drupal.conf [hunk 3]
```diff
@@ -1,5 +1,5 @@
...
     root /var/www/html;
```
Comment (zaporylie) [inline mainline-alpine/drupal.conf]: IMHO we should tokenize `root` as well. I suggest name nginx independent so I could implement it the same way in drupal-docker/apache (maybe `DOCROOT`).

```
root DOCROOT
```

Probably even better to use before/after special char, something like `%SERVER_NAME%` and `%DOCROOT%`

File: mainline-alpine/entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#!/bin/sh
+
+set -e
+
+# Configure docroot.
+if [ -n "$NGINX_DOCROOT" ]; then
```
Comment (zaporylie) [inline mainline-alpine/entrypoint.sh]: As I have mentioned before, this variable should be nginx independent. Maybe just `DOCROOT`?

File: mainline-alpine/entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+# Ensure server name defined.
+if [ -z "$NGINX_SERVER_NAME" ]; then
+    NGINX_SERVER_NAME=drupal
+fi
+
+# Set server name
+sed -i 's/SERVER_NAME/'"${NGINX_SERVER_NAME}"'/' /etc/nginx/conf.d/*.conf
+echo "Configured nginx server name to ${NGINX_SERVER_NAME}."
+
+exec nginx -g "daemon off;"
```
Comment (zaporylie) [inline mainline-alpine/entrypoint.sh]: This script is suppose to be `ENTRYPOINT` so it should accept command argument and execute it. See https://github.com/docker-library/mysql/blob/master/5.5/docker-entrypoint.sh#L151 or https://github.com/docker-library/redis/blob/master/3.0/docker-entrypoint.sh#L16

File: mainline-alpine/Dockerfile [hunk 6]
```diff
@@ -11,3 +11,6 @@ RUN addgroup -g 82 -S www-data \
 
 COPY nginx.conf /etc/nginx/nginx.conf
 COPY drupal.conf /etc/nginx/conf.d/default.conf
+
```
Comment (zaporylie) [inline mainline-alpine/Dockerfile:14]: Environment variables should be declared in Dockerfile.

**Review by zaporylie — CHANGES_REQUESTED** [review]
I like sed-based approach. Have some doubts though about naming convention and entrypoint usage. Added some inline comments to place it in the context.

Just created pull request to your fork-branch with my suggestions. Curious to hear your opinion :)

Thank you for your involvement and pushing this long-standing issue forward.

**Comment by slashrsm** [comment]
Makes sense. It your PR you didn't use replacement tokens with special characters as you mentioned in one of the comments. Should we do that?

Thank you for helping. I am constantly learning when contributing to this project.

**Comment by zaporylie** [comment]
Yes, I think we should. As we use `sed` to change _all_ the config files containing SERVER_NAME or DOCROOT (not just default.conf) we should make sure we replace only tokens which suppose to be replaced, not all occurrences of SERVER_NAME and DOCROOT string.

Still not sure why you suggested changing default server_name from `_` to `drupal`? Probably both would work by default but IMHO `_` should be used for catch-all server as it is quite common practice (ref. http://nginx.org/en/docs/http/server_names.html).

**Comment by slashrsm** [comment]
> Yes, I think we should. As we use sed to change all the config files containing SERVER_NAME or DOCROOT (not just default.conf) we should make sure we replace only tokens which suppose to be replaced, not all occurrences of SERVER_NAME and DOCROOT string.

Done

> Still not sure why you suggested changing default server_name from _ to drupal? Probably both would work by default but IMHO _ should be used for catch-all server as it is quite common practice (ref. http://nginx.org/en/docs/http/server_names.html).

Not really sure about that. Reverted back to "_" while still offering to override it.

**Review by zaporylie — APPROVED** [review]

**Pull request merged**